crates/astria-sequencer/src/service/mempool/mod.rs

```diff
@@ -119,50 +128,159 @@ async fn handle_check_tx<S: accounts::StateReadExt + address::StateReadExt + 'st
 ) -> response::CheckTx {
     use sha2::Digest as _;
 
-    let start_parsing = Instant::now();
-
     let request::CheckTx {
         tx, ..
     } = req;
 
     let tx_hash = sha2::Sha256::digest(&tx).into();
+
+    // check if the transaction has been removed from the appside mempool
+    if let Err(rsp) = check_removed_comet_bft(tx_hash, mempool, metrics).await {
+        return rsp;
+    }
+
+    // check if the transaction is already in the mempool
+    if is_tracked(tx_hash, mempool, metrics).await {
+        return response::CheckTx::default();
+    }
+
+    // perform stateless checks
+    let signed_tx = match stateless_checks(tx, &state, metrics).await {
+        Ok(signed_tx) => signed_tx,
+        Err(rsp) => return rsp,
+    };
+
+    // attempt to insert the transaction into the mempool
+    if let Err(rsp) = insert_into_mempool(mempool, &state, signed_tx, metrics).await {
+        return rsp;
+    }
+
+    // insertion successful
+    metrics.set_transactions_in_mempool_total(mempool.len().await);
+
+    response::CheckTx::default()
+}
+
+/// Checks if the transaction is already in the mempool.
+async fn is_tracked(tx_hash: [u8; 32], mempool: &AppMempool, metrics: &Metrics) -> bool {
+    let start_tracked_check = Instant::now();
+
+    let result = mempool.is_tracked(tx_hash).await;
+
+    metrics.record_check_tx_duration_seconds_check_tracked(start_tracked_check.elapsed());
+
+    result
+}
+
+/// Checks if the transaction has been removed from the appside mempool.
+///
+/// Returns an `Err(response::CheckTx)` with an error code and message if the transaction has been
+/// removed from the appside mempool.
+async fn check_removed_comet_bft(
+    tx_hash: [u8; 32],
+    mempool: &AppMempool,
+    metrics: &Metrics,
+) -> Result<(), response::CheckTx> {
```

Comment:
You could just return a `RemovalReason` here and instead implement a new trait `IntoCheckTxResponse` on it so that you don't need to do the conversion inline.